gap/Tools.gi

```diff
@@ -0,0 +1,196 @@
+                e[1] := e[1] - 1;
+                
+            fi;
+            
+            Assert( 0, e[2] <> pos - 1 );
+            
+            if e[2] > pos - 1 then
+                
+                e[2] := e[2] - 1;
+                
+            fi;
+            
+            return e;
+            
+        end );
+        
+    od;
+    
+    # See https://github.com/Quantomatic/quantomatic/blob/stable/docs/json_formats.txt
+    # for an overview of the qgraph format.
+    
+    wire_vertices := rec( );
+    node_vertices := rec( );
+    
+    for pos in [ 1 .. Length( labels ) ] do
+    
+    
+        if labels[ pos ] = "X" then
+        
+            node_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                              data := rec( type := "X" ) );
+        
+        elif labels[ pos ] = "Z" then
+        
+            node_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                              data := rec( type := "Z" ) );
+        
+        elif labels[ pos ] = "H" then
+        
+            node_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                              data := rec( type := "hadamard" ) );
+                                              # TODO: when to set is_edge := true?
+        
+        elif labels[ pos ] = "input" then
+        
+           wire_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                             input := true,
+                                             output := false );
+        
+        elif labels[ pos ] = "output" then
+        
+           wire_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                             input := false,
+                                             output := true );
+        
+        elif labels[ pos ] = "input_output" then
+        
+           wire_vertices.( pos - 1 ) := rec( annotation := rec( coord := [ 0, 0 ] ),
+                                             input := true,
+                                             output := true );
+        
+        # TODO: neutral nodes?
+        fi;
+        
+    od;
+    
+    undir_edges := rec( );
+    
+    for edge in edges do
+    
+        source := edge[1];
+        target := edge[2];
+        
+        edge_name := Concatenation( "E", String( source ), String( target ) );
+        
+        undir_edges.( edge_name ) := rec( src := String( source ), tgt := String( target ) );
+    
+    od;
+    
+    qgraph := rec( wire_vertices := wire_vertices,
+                   node_vertices := node_vertices,
+                   undir_edges := undir_edges );
+    
+    qgraph := GapToJsonString( qgraph );
+    
+    file := OutputTextFile( Concatenation( filename, ".qgraph" ), false );;
+    
+    SetPrintFormattingStatus( file, false );
+    
+    PrintTo( file, qgraph );
+    
```

Comment:
Please use `CloseStream` to close the output stream here (see documentation of `OutputTextFile`).

Reply:
Or use `FileString` which should be more convenient anyway :-) (Yes, GAP and homalg have lots of different IO strategies :D )

Reply:
Thank you, yes, that is confusing (and not so easy to find). :D
Also done.